File: swift/planet-time/Sources/InterplanetTime/Scheduling.swift
// Scheduling.swift — Meeting window finder ported from planet-time.js.

import Foundation

public extension InterplanetTime {

    struct MeetingWindow {
        public let startMs, endMs: Int64
        public let durationMinutes: Int
    }

    static func findMeetingWindows(
        _ a: String, _ b: String,
        from fromMs: Int64,
        earthDays: Int = 7,
        stepMin: Int = 15
    ) -> [MeetingWindow] {
        let stepMs = Int64(stepMin) * 60_000
        let endMs  = fromMs + Int64(earthDays) * earthDayMs
        var windows: [MeetingWindow] = []
        var inWindow = false
        var winStart = Int64(0)

        var t = fromMs
        while t < endMs {
            let ptA = getPlanetTime(a, t)
            let ptB = getPlanetTime(b, t)
            let both = ptA.isWorkHour && ptB.isWorkHour

            if both && !inWindow {
                inWindow = true; winStart = t
            } else if !both && inWindow {
                inWindow = false
                let dur = Int((t - winStart) / 60_000)
                if dur > 0 { windows.append(MeetingWindow(startMs: winStart, endMs: t, durationMinutes: dur)) }
            }
            t += stepMs
        }
        if inWindow {
            let dur = Int((endMs - winStart) / 60_000)
            if dur > 0 { windows.append(MeetingWindow(startMs: winStart, endMs: endMs, durationMinutes: dur)) }
        }
        return windows
    }
}

File: scala/planet-time/src/main/scala/interplanet/time/Scheduling.scala
package interplanet.time

/**
 * Scheduling.scala — findMeetingWindows
 * Ported from planet-time.js v1.1.0
 */

/**
 * Find overlapping work windows between two planets over N Earth days.
 * Step size: 15 minutes.
 */
def findMeetingWindows(
  planetA: Planet,
  planetB: Planet,
  earthDays: Int = 7,
  startMs: Long = System.currentTimeMillis()
): List[MeetingWindow] =
  val stepMs = 15 * 60000L
  val endMs = startMs + earthDays * EARTH_DAY_MS
  val windows = scala.collection.mutable.ListBuffer[MeetingWindow]()
  var inWindow = false
  var windowStart = 0L

  var t = startMs
  while t < endMs do
    val ta = getPlanetTime(planetA, t)
    val tb = getPlanetTime(planetB, t)
    val overlap = ta.isWorkHour && tb.isWorkHour
    if overlap && !inWindow then
      inWindow = true
      windowStart = t
    if !overlap && inWindow then
      inWindow = false
      windows += MeetingWindow(
        startMs = windowStart,
        endMs = t,
        durationMin = ((t - windowStart) / 60000L).toInt
      )
    t += stepMs

  if inWindow then
    windows += MeetingWindow(
      startMs = windowStart,
      endMs = endMs,
      durationMin = ((endMs - windowStart) / 60000L).toInt
    )

  windows.toList
